doc/lock_files.md

```diff
@@ -0,0 +1,239 @@
+      sha256: 83de9017d911cf7795f19353dba4d04bd24cd40622b7567ff61fc3f7223aa3ea
+
+compiler: ghc-X.Y.Z
+
+packages:
+  acme-missiles:
+    location:
+      # QUESTION: any reason we need to specify which snapshot file it came from? I don't think so...
+      original: https://hackage.haskell.org/package/acme-missiles-0.3.tar.gz
+      completed:
+        size: 1442
+        url: https://hackage.haskell.org/package/acme-missiles-0.3.tar.gz
+        cabal-file:
+          size: 613
+          sha256: 2ba66a092a32593880a87fb00f3213762d7bca65a687d45965778deb8694c5d1
+        name: acme-missiles
+        version: '0.3'
+        sha256: e563d8b524017a06b32768c4db8eff1f822f3fb22a90320b7e414402647b735b
+        pantry-tree:
+          size: 226
+          sha256: 614bc0cca76937507ea0a5ccc17a504c997ce458d7f2f9e43b15a10c8eaeb033
+    flags: ...
+    hidden: true/false
+    ghc-options: [...]
+```
+
+**NOTE** The `original` fields may seem superfluous at first. See the
+update procedure below for an explanation.
+
+## Creation
+
+Whenever a `stack.yaml` file is loaded, Stack checks for a lock file
+in the same file path, with a `.lock` extension added. For example, if
+you run `stack build --stack-yaml stack-11.yaml`, it will use a lock
+file in the location `stack-11.yaml.lock`. For the rest of this
+document, we'll assume that the files are simply `stack.yaml` and
+`stack.yaml.lock`.
+
+If the lock file does not exist, it will be created by:
+
+* Loading the `stack.yaml`
+* Loading all snapshot files
+* Completing all missing information
+* Writing out the new `stack.yaml.lock` file
+
+## Dirtiness checking
+
+If the `stack.yaml.lock` file exists, its last modification time is
+compared against the last modification time of the `stack.yaml` file
+and any local snapshot files. If any of those files is more recent
+than the `stack.yaml` file, then the update procedure is
+triggered. Otherwise, the `stack.yaml.lock` file can be used as the
+definition of the snapshot.
+
+## Update procedure
+
+The simplest possible implementation is: ignore the lock file entirely
+and create a new one followign the creation steps above. There's a
+significant downside to this, however: it may cause a larger delta in
+the lock file than intended, by causing more packages to be
+updates. For example, many packages from Hackage may have their
+Hackage revision information updated unnecessarily.
+
+The more complicated update procedure is described below. **QUESTION**
+Do we want to go the easy way at first and later implement the more
+complicated update procedure?
+
+1. Create a map from original package location to completed package
+   location in the lock file
+2. Load up each snapshot file
+3. For each incomplete package location:
+    * Lookup the value in the map created in (1)
+    * If present: use that completed information
+    * Otherwise: complete the information using the procedure in
+      "creation"
```

Comment:
The only problem is that "creation" doesn't tell much about the procedure

Reply:
If there's anything ambiguous about the procedure, please ask. I thought the procedure is fairly straightforward based on existing code we have in both Stack and Pantry.

Reply:
I think it could be a bit confusing for a person not quite familiar with the details in Pantry - maybe it should say something like "using the same completion procedure from Pantry as in "creation""?

Reply:
Good call, updated.